### DISCUSSION TITLE
RBAC, ABAC, and GBAC for the GraphQL API

### DISCUSSION THREAD

**Discussion by amaster507** [geldata/gel#3416] [Pre-RFC]
I am eyeing replacements for my Dgraph Cloud backend. Haven't reached a verdict yet though whether to ride out the storm or jump ship.

What I need is a mature GraphQL API that is capable of serving our highly connected data set to many users across the globe who are highly focused on the security of their data.

We need three types of authorization rules to allow clients of the API to not just be able to do functions such as create/read/update/delete but also use properties of the nodes in the graph to grant authorization to CRUD actions for only some of the nodes. This would allow us to serve users the same API and then users can access their data, the public data, or their group's data while not being able to access other users data or non-public data.

In Dgraph Cloud, I can write auth rules in GraphQL that traverse the graph and grant access to the data depending on properties of the connected data. Or for some types, I just need role based access such as only allowing administrators access to create and delete users.

I need RBAC (Role Based Access Control), ABAC (Attribute Based Access Control), and GBAC (Graph Based Access Control) rules.

Here is a GraphQL schema for example:
```
type Contact {
  id: ID!
  public: Boolean
  access: [ACL]
}
type Task {
  id: ID!
  name: String!
  due: Datetime
  comp: Datetime
  owner: User
}
type User {
  id: ID!
  username: String!
  email: String!
  tasks: [Task]
  role: [String!]!
}
type Group {
  id: ID!
  name: String!
  members: [User]
}
type ACL {
  id: ID
  level: String!
  owner: Owner
}
union Owner = User | Group
```

Here are some of the auth rules I need in my GraphQL API:
- All authenticated users can see all public Contacts
- Users who have the administrator role can see all Contacts
- Users can see Contacts they own connected by access ACL type
- Users can see Contacts that are owned by a Group connected by access ACL types that they are a member of
- Users can see their own Tasks
- Users can only see Groups they are a member of

And these are just some of the rules for read. Other rules will be needed for create, update, and delete.

Will EdgeDB ever be capable of generating a GraphQL API with such advanced rule based authorization control of the data.

And even better would be if data could be authorized by fields as well. So maybe I want to allow Users to see other users, but only their own User's email address field. Same kind of logic for create, update, and delete...

Anyone's thoughts or feedback on this?

**Comment by elprans** [comment]
Hi @amaster507!

We are working on the formal specification of access control policies in EdgeDB schemas.  There is a draft RFC in https://github.com/edgedb/rfcs/pull/50, although it will likely be superseded by a lighter RFC soon.  To answer your question more specifically: yes, all of the above use cases are planned to be supported.

**Reply by pheuter** [reply]
@elprans My team has just begun a major rewrite of our platform, migrating from MongoDB over to Postgres + Hasura. Needless to say, EdgeDB caught my attention and seems like a fantastic solution for us as an alternative to Hasura. The current biggest blocker is support for ACL/Authorization.

I noticed the Roadmap mentions it's coming in "1.X", but I was hoping you'd be able to provide a rough timeline for when we could expect that as it would influence our decision making. Within 6 months? A year? Any additional insight would be appreciated.
